{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "WBT-MBT2-25E <i>Programming Python for Bioinformatics</i> &copy; 2020-2023 Michal Bukowski (m.bukowski@uj.edu.pl) Department of Analytical Biochemistry, Faculty of Biochemistry, Biophysics and Biotechnology, Jagiellonian University\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddeeff;border-radius:15px\">DNASeq, a new data type for processing DNA sequences</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=3>\n",
    "    \n",
    "Below you will find an incomplete implementation of the class `DNASeq` that is supposed to define a new type of a data structure facilitating different operations performed on DNA sequences. These include operations that might be used to process multiple sequence alignments in FASTA format, which apart from letters belonging to the [IUPAC ambiguous DNA alphabet](https://www.bioinformatics.org/sms/iupac.html), may contain gaps (the minus sign `-`).<br>\n",
    "\n",
    "Complete gradually the implentation fo the class DNASeq (<u>it must be contained in one cell</u>). Follow the comments in the code. Upon completion of each numbered part, run code in the last section `Test the code!`. It will help you to build the class step by step.<br>\n",
    "\n",
    "<u>Remove the comment sign</u> `#` placed before subsequent code lines (manually or using `Ctrl + /` for multiple lines at once) and fill in the gaps marked with `???`. Sometimes it requires only one line of code or a simple expression, sometimes a whole block of code.<br>\n",
    "\n",
    "Again, remember to fill the code step by step, part by part.<br>\n",
    "    \n",
    "<b>REMEMBER</b> to <u>rerun</u> the cell with DNASeq class definition after making any changes. Otherwise, those will not take any effect.<br>\n",
    "    \n",
    "<b>TEST</b> your code upon finishing each section. It <u>must work</u> before you move forward. In order to make the code work, you <u>must implement sections in the given order</u>.\n",
    "\n",
    "</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### On the GenBank notation\n",
    "#### <u>An example for positions/coordinates/indices 1 and 4</u>:\n",
    "(<i>start</i> and <i>stop</i> in Python, <i>begining</i> and <i>end</i> in GenBank)\n",
    "<pre style=\"font-family:Monospace\">\n",
    "<span style=\"color:#00aaaa\">          |----&gt;|</span>\n",
    "<span style=\"color:#aaaaaa\"> Python: 0 1 2 3 4 5 6 7 8</span>\n",
    "         A T C G T A C G A\n",
    "<span style=\"color:#aaaaaa\">GenBank: 1 2 3 4 5 6 7 8 9</span>\n",
    "<span style=\"color:#00aaaa\">        |------&gt;|</span>\n",
    "         T A G C A T G C T   (complementary/minus strand)\n",
    "<span style=\"color:#aa0000\">        |&lt;------|</span>\n",
    "\n",
    "<span style=\"color:#aaaaaa\"> Python:</span> seq[5]   <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'A'</span>\n",
    "<span style=\"color:#aaaaaa\">GenBank:</span> seq[5]   <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'T'</span>\n",
    "\n",
    "<span style=\"color:#aaaaaa\"> Python:</span> seq[1:4] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'TCG'</span>\n",
    "<span style=\"color:#aaaaaa\">GenBank:</span> seq[1:4] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'ATCG'</span>\n",
    "\n",
    "<span style=\"color:#aaaaaa\"> Python:</span> seq[4:1] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#aa0000\">''</span>\n",
    "<span style=\"color:#aaaaaa\">GenBank:</span> seq[4:1] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#aa0000\">'CGAT'</span>   (5' to 3' direction, reverse-complement)\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The overview of the implemented class\n",
    "\n",
    "    class DNASeq:\n",
    "    \n",
    "        ALPH\n",
    "        FASTA_REGEX\n",
    "        \n",
    "        __init__()\n",
    "        __repr__()\n",
    "        __str__()\n",
    "        __len__()\n",
    "        __add__()\n",
    "        __getitem__()\n",
    "        \n",
    "        from_file()\n",
    "        revcmpl()\n",
    "        copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# and there it is, our DNASeq class!\n",
    "class DNASeq:\n",
    " \n",
    "    #===SECTION=1========================================================\n",
    "    \n",
    "    # a dictionary with the IUPAC ambiguous DNA alphabet\n",
    "    # (https://www.bioinformatics.org/sms/iupac.html)\n",
    "    # and a gap sign, the key is paired with a value\n",
    "    # that is the key's complement, this dictionary\n",
    "    # is defined as a class variable,\n",
    "    # revise how to access values of such variables,\n",
    "    # TODO: fill up the missing letters,\n",
    "    #       make them capitals\n",
    "    \n",
    "    ALPH = {\n",
    "       'A' : 'T',   'T' : 'A',   'G' : 'C',   'C' : 'G',\n",
    "       'K' : 'M',   'M' : 'K',   'B' : 'V',   'D' : 'H',\n",
    "       'H' : 'D',   'V' : 'B',   'N' : 'N',   '-' : '-'\n",
    "    }\n",
    "    \n",
    "    # regular expression for FASTA format processing,\n",
    "    # another class variable,\n",
    "    # TODO: assign to FASTA_REGEX variable a proper\n",
    "    #       value to formulate a proper regular expression\n",
    "    \n",
    "    FASTA_REGEX = '\\>([^\\ \\n]+)\\ ?(.*)\\n([^\\>]+)'\n",
    "\n",
    "    \n",
    "    # TEST the code in the Section 1\n",
    "    \n",
    "    \n",
    "    #===SECTION=2========================================================\n",
    "    \n",
    "    # the special method __init__() initialising the\n",
    "    # initial state of a newly created object (a class instance),\n",
    "    # TODO: define the special method __init__() proper for\n",
    "    #       initialisation of a new object of DNASeq type,\n",
    "    #       which is to be created as following:\n",
    "    #       seq = DNASeq(seqid, title, seq)\n",
    "    \n",
    "    def __init__(self, seqid, title, seq):\n",
    "        # TODO: assign values of the arguments passed\n",
    "        #       whilst creating a new object to the\n",
    "        #       object attributes of the same name\n",
    "        \n",
    "        self.seqid = seqid\n",
    "        self.title = title\n",
    "        self.seq = seq\n",
    "        \n",
    "    # TEST the code in the Section 2\n",
    "        \n",
    "        \n",
    "    #===SECTION=3========================================================\n",
    "        \n",
    "    # now we need a class method that will help us to\n",
    "    # deserialise objects from a file in FASTA format\n",
    "    # TODO: define a class method from_file that next to\n",
    "    #       the automatically passed reference to a class\n",
    "    #       will accept an argument with a file path, and\n",
    "    #       will work as following:\n",
    "    #       seqs = DNASeq.from_file('some/path/some_file.fasta')\n",
    "    \n",
    "    @classmethod\n",
    "    def from_file(cls, filename):\n",
    "        \n",
    "        seqs = {}      \n",
    "        \n",
    "        # TODO: load the content of the file to\n",
    "        #       a variable buf by using the\n",
    "        #       instruction with\n",
    "        with open(filename) as f: buffer = f.read()\n",
    "\n",
    "        \n",
    "        # TODO: start iteratively searching for FASTA_REGEX in\n",
    "        #       buf by using a for loop and the re.finditer() method        \n",
    "        for match in re.finditer(cls.FASTA_REGEX, buffer):\n",
    "            \n",
    "            seqid, title, seq = match.groups()\n",
    "            \n",
    "            seq = re.sub('[\\ \\n\\t]+', '', seq).upper()\n",
    "            \n",
    "            seq_alph = set(seq)\n",
    "              \n",
    "            # TODO: find a proper method for Python sets,\n",
    "            #       which will allow you to see if the\n",
    "            #       alphabet of seq is a subset of ALPHABET,\n",
    "            #       if not raise an exception            \n",
    "            if seq_alph in set(DNASeq.ALPH):\n",
    "                raise Exception(f'Sequence {seqid} contains incorrect letters')\n",
    "\n",
    "            seqs[seqid] = cls(seqid, title, seq)\n",
    "              \n",
    "        return seqs\n",
    "    \n",
    "    # TEST the code in the Section 3\n",
    "    \n",
    "        \n",
    "    #===SECTION=4========================================================\n",
    "    \n",
    "    # the special method __repr__() returns a string representation of\n",
    "    # an object, we will define this represetation as 10 first letters\n",
    "    # of the contained sequence and three dots '...' if the sequnce\n",
    "    # is longer than 10 letters\n",
    "    \n",
    "    def __repr__(self):\n",
    "        \n",
    "        # get sequence until 10th character\n",
    "        repr = self.seq[:10]\n",
    "        \n",
    "        # if the length of the sequence is greater to 10\n",
    "        if len(self.seq) > 10:\n",
    "            # add ... in the end\n",
    "            repr = repr+'...'\n",
    "        \n",
    "        return repr\n",
    "    \n",
    "    # TEST the code in the Section 4\n",
    "    \n",
    "    \n",
    "    #===SECTION=5========================================================\n",
    "        \n",
    "    # TODO: define the proper special method that will return the lenght of\n",
    "    #       the sequence contained within a given DNASeq object when the reference\n",
    "    #       to that object is passed to the built-in len() function\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.seq)\n",
    "    \n",
    "    # TEST the code in the Section 5\n",
    "    \n",
    "    \n",
    "    #===SECTION=6========================================================\n",
    "    \n",
    "    # TODO: implement the special method __str__(), which returns\n",
    "    #       a string whenever a DNASeq object is being converted to one,\n",
    "    #       eg. str(obj) or print(obj),\n",
    "    #       make it return the sequence contained in an object \n",
    "    #       of DNASeq type as FASTA format sequence string,\n",
    "    #       let the sequence be divided into 60-character lines\n",
    "    \n",
    "    def __str__(self):\n",
    "        \n",
    "        seqid = '>'+self.seqid\n",
    "        \n",
    "        lines = '\\n'.join(self.seq[i:i+60] for i in range(0, len(self.seq), 60))\n",
    "\n",
    "        title = f' {self.title}' if self.title != '' else ''\n",
    "\n",
    "        fasta = f'{seqid}{title}\\n{lines}'\n",
    "        \n",
    "        return fasta\n",
    "    \n",
    "    \n",
    "    # TEST the code in the Section 6\n",
    "    \n",
    "    \n",
    "    #===SECTION=7=======================================================\n",
    "    \n",
    "    # let's create a custom method and call it revcmpl(),\n",
    "    # it simply returns a new object of DNASeq type, which\n",
    "    # will contain a reverse-complement sequence to the\n",
    "    # one contained in the object the method is called from,\n",
    "    # to make both object distinguishable, create a new\n",
    "    # seqid for the sequence by adding '_revcmpl' suffix\n",
    "    # to the seqid of the original one\n",
    "    \n",
    "    def revcmpl(self):\n",
    "        # TODO:convert sequence contained in the object\n",
    "        #      to a list called seq\n",
    "        \n",
    "        seq = self.seq\n",
    "        \n",
    "        # TODO: reverse the list in-place        \n",
    "        inv = seq[::-1]\n",
    "       \n",
    "        # TODO: using string method join(), the class dictionary ALPH and a\n",
    "        #       list comprehension, translate the reversed sequence and\n",
    "        #       convert into a string\n",
    "        \n",
    "        seq_revcmpl = ''.join(self.ALPH[inv[i]] for i in range(len(inv)))\n",
    "        \n",
    "        # TODO: create seqid variable and assign to it the object's seqid\n",
    "        #       and the suffix '_revcmpl'\n",
    "        \n",
    "        seqid = self.seqid+'_revcmpl'\n",
    "        \n",
    "        # TODO: create a new object of the DNASeq type using the new seqid,\n",
    "        #       title contained in the object as well as\n",
    "        #       reveresed and translated sequence, return the new object\n",
    "        \n",
    "        revcmpl_dnaseq = DNASeq(seqid, self.title, seq_revcmpl)\n",
    "        \n",
    "        return revcmpl_dnaseq\n",
    "    \n",
    "    # TEST the code in the Section 7\n",
    "    \n",
    "    \n",
    "    #===SECTION=8========================================================\n",
    "    \n",
    "    # beware! this is the hardest part to go through,\n",
    "    # we will implement the special method __getitem__()\n",
    "    # that allows to program what happens when an object\n",
    "    # is indexed or sliced (like a list or string),\n",
    "    # eg. obj[3] or obj[4:5]\n",
    "    #\n",
    "    # we know that indexing and slicing in Python works as follows:\n",
    "    # - indexing starts at 0 and goes up\n",
    "    # - when a slice is taken [3:4] the first index is included\n",
    "    #   the last not: eg. s = 'abcde', s[1:3] -> 'bc'\n",
    "    #                          01234\n",
    "    # - slice [3:1] will be an empty one\n",
    "    #\n",
    "    # when it comes to biological sequences, those are indexed\n",
    "    # according to GenBank notation, which is quite different:\n",
    "    # - indexing starts at 1\n",
    "    # - when a fragment is requested, both indices are inclusive:\n",
    "    #   seq = 'ATGCTACG', seq[1:3] -> 'ATG'\n",
    "    #          12345678\n",
    "    # - the start index greater than stop index indicates\n",
    "    #   a reverse complement (the complementary strand):\n",
    "    #   seq[3:1] -> 'CAT'\n",
    "    #\n",
    "    # now we will try to implement this behaviour in case of\n",
    "    # our objects of DNASeq type\n",
    "    \n",
    "    def __getitem__(self, key):\n",
    "        \n",
    "        if isinstance(key, slice):\n",
    "            # if the key is a slice object, it has three properties:\n",
    "            # start, stop and step, eg. list[start:stop:step]\n",
    "            # if any is missing its value is None, eg. list[start:end]\n",
    "            # first two will be equivalent of beginning and end\n",
    "            # in GenBank notation\n",
    "            \n",
    "            # the whole point here is to translate GenBank indices\n",
    "            # into Python ones and index or slice the sequence\n",
    "            # contained in the DNASeq object, which is surely of\n",
    "            # Python string format, so it cannot be directly indexed\n",
    "            # or sliced with GenBank indices wihtout translating them\n",
    "            \n",
    "            # to make the code nicer, let's assign values of slice\n",
    "            # properties to single variables\n",
    "            \n",
    "            start, end, step = key.start, key.stop, key.step\n",
    "            \n",
    "            # let's test a few possibilites and exclude those\n",
    "            # that cannot be translated into GenBank notation\n",
    "            \n",
    "            if step is not None:\n",
    "                # there is not an equivalent of step in GenBank notation\n",
    "                # raise a KeyError                \n",
    "                raise KeyError('Step is not allowed in GenBank notation')\n",
    "            \n",
    "            if start is None:\n",
    "                # start must be provided                \n",
    "                raise KeyError('start index is required')\n",
    "            \n",
    "            if end is None:\n",
    "                # as well as end                \n",
    "                raise KeyError('end index is required')\n",
    "\n",
    "                \n",
    "            if not np.issubdtype(type(start), np.integer) or \\\n",
    "              not np.issubdtype(type(end), np.integer):\n",
    "                # start and end must be defined as integer values,\n",
    "                # otherwise raise a KeyError\n",
    "                \n",
    "                raise TypeError('Start and end must be integers')\n",
    "            \n",
    "            if start <= 0 or end <= 0:\n",
    "                # both must be greater than 0 as in GenBank notation,\n",
    "                # in which indices starts at 1                \n",
    "                raise KeyError('Minimal value for start and end is 1')\n",
    "                \n",
    "            # now we are sure there is only start and end in the slice,\n",
    "            # and that both are integer type values and equal or greater to 1,\n",
    "            # let's move on then\n",
    "            \n",
    "            # TODO: create a variable strand and set it to 1 if start is\n",
    "            #       less or equal to end, otherwise to -1\n",
    "            # TODO: if strand is equal to -1, swap the values of start and end,\n",
    "            #       by using unpacking, so start is less than end (they are ordered)\n",
    "            \n",
    "            if start <= end:\n",
    "                strand = 1\n",
    "            else:\n",
    "                strand = -1    \n",
    "                start, end = end, start\n",
    "            \n",
    "            # TODO: create a new seqid by adding to the existing one\n",
    "            #       suffix '_loc(start_end)', where start and end are\n",
    "            #       values of our variables            \n",
    "          \n",
    "            new_seqid = self.seqid+f'_loc({start}_{end})'\n",
    "            \n",
    "            # TODO: decrease start by 1, if start in GenBank notation is 1,\n",
    "            #       it is and equivalent of 0 in Python indexing (etc.),\n",
    "            #       so it needs to be decreased by 1 to be translated\n",
    "            #       from GenBank to Python\n",
    "            #\n",
    "            #       you must not decrease end, as in GenBank it is inclusive,\n",
    "            #       but in Python exclusive, so it is decreased\n",
    "            #       somewhat automatically\n",
    "            \n",
    "            start -= 1 \n",
    "            \n",
    "            # TODO: create a new object of DNASeq type, by using new seqid\n",
    "            #       you created a few lines before, title of the existing object,\n",
    "            #       a slice of the sequence contained in that object by using\n",
    "            #       the adjusted start index and the end index (string slicing)    \n",
    "            \n",
    "            title = self.title\n",
    "            sub_seq = self.seq[start:end]\n",
    "            new_seq = DNASeq(new_seqid, title, sub_seq)\n",
    "\n",
    "            \n",
    "            # TODO: if strand is -1, assign to the same reference (\"variable\")\n",
    "            #       a reverse complement of the new sequence object by using\n",
    "            #       the method revcmpl() you implemented before\n",
    "            \n",
    "            if strand == -1:\n",
    "                new_seq = new_seq.revcmpl()        \n",
    "                                \n",
    "            # TODO: return the new sequence object       \n",
    "            \n",
    "            return new_seq\n",
    "                \n",
    "        else:\n",
    "            \n",
    "            # if we are here, it means that key is not a slice object,\n",
    "            # then we allow it be an integer greather than 0,\n",
    "            # which is GenBank notation\n",
    "            \n",
    "            if not np.issubdtype(type(key), np.integer):\n",
    "                raise TypeError('Index must be an integer')\n",
    "                \n",
    "            if key <= 0:\n",
    "                raise KeyError('Minimal value of index is 1')\n",
    "                \n",
    "            # in case it is just one letter not a slice, we\n",
    "            # will return simply a letter from the sequence,\n",
    "            # not a new DNASeq object, we need to remeber to\n",
    "            # decrease the key value by one to translate it from\n",
    "            # GenBank to Python\n",
    "                \n",
    "            return self.seq[key-1]\n",
    "    \n",
    "    # TEST the code in the Section 8\n",
    "        \n",
    "        \n",
    "    #===SECTION=9========================================================\n",
    "    \n",
    "    # implement the special method __add__() which will be\n",
    "    # invoked when two objects of DNASeq type are being added,\n",
    "    # eg. seq_3 = seq_1 + seq_2\n",
    "    #\n",
    "    # create the result of addition as a new object of DNASeq type\n",
    "    # and set its:\n",
    "    # - seqid to seqids of added objects separated by an underscore\n",
    "    # - title to titles of added objects separated by an underscore\n",
    "    # - seq to a simple concatenation of both sequences\n",
    "    #\n",
    "    # return the new object\n",
    "    \n",
    "    def __add__(seq1, seq2):\n",
    "        add_seqid = seq1.seqid+f'_{seq2.seqid}'\n",
    "        add_title = seq1.title+f'_{seq2.title}'\n",
    "        add_seq = seq1.seq + seq2.seq\n",
    "        \n",
    "        add_dnaseq = DNASeq(add_seqid, add_title, add_seq)\n",
    "        \n",
    "        return add_dnaseq\n",
    "    \n",
    "    # TEST the code in the Section 9\n",
    "    \n",
    "    \n",
    "    #===SECTION=10========================================================\n",
    "        \n",
    "    # define a custom method copy() that will return an exact copy\n",
    "    # of the existing object\n",
    "        \n",
    "    def copy(self):\n",
    "        \n",
    "        seqid = self.seqid\n",
    "        title = self.title\n",
    "        seq = self.seq\n",
    "        \n",
    "        copy_dnaseq = DNASeq(seqid, title, seq)\n",
    "        \n",
    "        return copy_dnaseq\n",
    "        \n",
    "    \n",
    "    # TEST the code in the Section 10\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddeeff;border-radius:15px\">Test the code!</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 1<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "The whole dictionary: {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'K': 'M', 'M': 'K', 'B': 'V', 'D': 'H', 'H': 'D', 'V': 'B', 'N': 'N', '-': '-'}<br><br>\n",
    "A letter complementary to \"C\": G<br><br>\n",
    "Regular expression for FASTA format '\\\\>([^\\\\ \\n]+)\\\\ ?(.*)\\n([^\\\\>]+)'\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The whole dictionary: {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'K': 'M', 'M': 'K', 'B': 'V', 'D': 'H', 'H': 'D', 'V': 'B', 'N': 'N', '-': '-'} \n",
      "\n",
      "A letter complementary to \"C\": G \n",
      "\n",
      "Regular expression for FASTA format '\\\\>([^\\\\ \\n]+)\\\\ ?(.*)\\n([^\\\\>]+)' \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# let's see our class variables:\n",
    "print('The whole dictionary:', DNASeq.ALPH, '\\n')\n",
    "\n",
    "print('A letter complementary to \"C\":', DNASeq.ALPH['C'], '\\n')\n",
    "\n",
    "print('Regular expression for FASTA format', repr(DNASeq.FASTA_REGEX), '\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 2<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "new_seq (an exemplary sequence): ATCGT...\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "new_seq (an exemplary sequence): ATCGT...\n"
     ]
    }
   ],
   "source": [
    "seq = DNASeq('new_seq', 'an exemplary sequence',\n",
    "             'ATCGTAGGATCGGATTAGAGCGATTAGCTAG')\n",
    "\n",
    "print(f'{seq.seqid} ({seq.title}): {seq.seq[:5]}...')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 3<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "arcC, aroE, glpF, gmk, pta, tpi, yqiL<br><br>\n",
    "yqiL (Acetyle coenzyme A acetyltransferase): GCGTT...\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arcC, aroE, glpF, gmk, pta, tpi, yqiL\n",
      "\n",
      "yqiL (Acetyle coenzyme A acetyltransferase): GCGTT...\n"
     ]
    }
   ],
   "source": [
    "# deserialise sequences from a FASTA format file\n",
    "# into a collection of objects of DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "seqids = ', '.join( seqs.keys() )\n",
    "\n",
    "print(seqids)\n",
    "\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "print(f'\\n{seq.seqid} ({seq.title}): {seq.seq[:5]}...')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 4<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "{'arcC': TTATTAATCC...,<br>\n",
    " 'aroE': AATTTTAATT...,<br>\n",
    " 'glpF': GGTGCTGATT...,<br>\n",
    " 'gmk': CGAATATTTG...,<br>\n",
    " 'pta': GCAACACAAT...,<br>\n",
    " 'tpi': CACGAAACAG...,<br>\n",
    " 'yqiL': GCGTTTAAAG...}\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'arcC': TTATTAATCC...,\n",
       " 'aroE': AATTTTAATT...,\n",
       " 'glpF': GGTGCTGATT...,\n",
       " 'gmk': CGAATATTTG...,\n",
       " 'pta': GCAACACAAT...,\n",
       " 'tpi': CACGAAACAG...,\n",
       " 'yqiL': GCGTTTAAAG...}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "seqs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 5<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "516\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "516"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# select one of the sequences by its sequence id (seqid)\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "# look up the length of the contained sequence\n",
    "len(seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 6<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;yqiL Acetyle coenzyme A acetyltransferase<br>\n",
    "GCGTTTAAAGACGTGCCAGCCTATGATTTAGGTGCGACTTTAATAGAACATATTATTAAA<br>\n",
    "GAGACGGGTTTGAATCCAAGTGAGATTGATGAAGTTATCATCGGTAACGTACTACAAGCA<br>\n",
    "GGACAAGGACAAAATCCAGCACGAATTGCTGCTATGAAAGGTGGCTTGCCAGAAACAGTA<br>\n",
    "CCTGCATTTACAGTGAATAAAGTATGTGGTTCTGGGTTAAAGTCGATTCAATTAGCATAT<br>\n",
    "CAATCTATTGTGACTGGTGAAAATGACATCGTGCTAGCTGGCGGTATGGAGAATATGTCT<br>\n",
    "CAGTCACCAATGCTTGTCAACAACAGTCGCTTCGGTTTTAAAATGGGACATCAATCAATG<br>\n",
    "GTTGATAGCATGGTATATGATGGTTTAACAGATGTATTTAATCAATATCATATGGGTATT<br>\n",
    "ACTGCTGAAAATTTAGTGGAGCAATATGGTATTTCAAGAGAAGAACAAGATACATTTGCT<br>\n",
    "GTAAACTCACAACAAAAAGCAGTACGTGCACAGCAA\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">yqiL Acetyle coenzyme A acetyltransferase\n",
      "GCGTTTAAAGACGTGCCAGCCTATGATTTAGGTGCGACTTTAATAGAACATATTATTAAA\n",
      "GAGACGGGTTTGAATCCAAGTGAGATTGATGAAGTTATCATCGGTAACGTACTACAAGCA\n",
      "GGACAAGGACAAAATCCAGCACGAATTGCTGCTATGAAAGGTGGCTTGCCAGAAACAGTA\n",
      "CCTGCATTTACAGTGAATAAAGTATGTGGTTCTGGGTTAAAGTCGATTCAATTAGCATAT\n",
      "CAATCTATTGTGACTGGTGAAAATGACATCGTGCTAGCTGGCGGTATGGAGAATATGTCT\n",
      "CAGTCACCAATGCTTGTCAACAACAGTCGCTTCGGTTTTAAAATGGGACATCAATCAATG\n",
      "GTTGATAGCATGGTATATGATGGTTTAACAGATGTATTTAATCAATATCATATGGGTATT\n",
      "ACTGCTGAAAATTTAGTGGAGCAATATGGTATTTCAAGAGAAGAACAAGATACATTTGCT\n",
      "GTAAACTCACAACAAAAAGCAGTACGTGCACAGCAA\n"
     ]
    }
   ],
   "source": [
    "# reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# select one of the sequences by its sequence id (seqid)\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "print(seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 7<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;yqiL_revcmpl Acetyle coenzyme A acetyltransferase<br>\n",
    "TTGCTGTGCACGTACTGCTTTTTGTTGTGAGTTTACAGCAAATGTATCTTGTTCTTCTCT<br>\n",
    "TGAAATACCATATTGCTCCACTAAATTTTCAGCAGTAATACCCATATGATATTGATTAAA<br>\n",
    "TACATCTGTTAAACCATCATATACCATGCTATCAACCATTGATTGATGTCCCATTTTAAA<br>\n",
    "ACCGAAGCGACTGTTGTTGACAAGCATTGGTGACTGAGACATATTCTCCATACCGCCAGC<br>\n",
    "TAGCACGATGTCATTTTCACCAGTCACAATAGATTGATATGCTAATTGAATCGACTTTAA<br>\n",
    "CCCAGAACCACATACTTTATTCACTGTAAATGCAGGTACTGTTTCTGGCAAGCCACCTTT<br>\n",
    "CATAGCAGCAATTCGTGCTGGATTTTGTCCTTGTCCTGCTTGTAGTACGTTACCGATGAT<br>\n",
    "AACTTCATCAATCTCACTTGGATTCAAACCCGTCTCTTTAATAATATGTTCTATTAAAGT<br>\n",
    "CGCACCTAAATCATAGGCTGGCACGTCTTTAAACGC\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">yqiL_revcmpl Acetyle coenzyme A acetyltransferase\n",
      "TTGCTGTGCACGTACTGCTTTTTGTTGTGAGTTTACAGCAAATGTATCTTGTTCTTCTCT\n",
      "TGAAATACCATATTGCTCCACTAAATTTTCAGCAGTAATACCCATATGATATTGATTAAA\n",
      "TACATCTGTTAAACCATCATATACCATGCTATCAACCATTGATTGATGTCCCATTTTAAA\n",
      "ACCGAAGCGACTGTTGTTGACAAGCATTGGTGACTGAGACATATTCTCCATACCGCCAGC\n",
      "TAGCACGATGTCATTTTCACCAGTCACAATAGATTGATATGCTAATTGAATCGACTTTAA\n",
      "CCCAGAACCACATACTTTATTCACTGTAAATGCAGGTACTGTTTCTGGCAAGCCACCTTT\n",
      "CATAGCAGCAATTCGTGCTGGATTTTGTCCTTGTCCTGCTTGTAGTACGTTACCGATGAT\n",
      "AACTTCATCAATCTCACTTGGATTCAAACCCGTCTCTTTAATAATATGTTCTATTAAAGT\n",
      "CGCACCTAAATCATAGGCTGGCACGTCTTTAAACGC\n"
     ]
    }
   ],
   "source": [
    "# reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# select one of the sequences by its sequence id (seqid)\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "new_seq = seq.revcmpl()\n",
    "\n",
    "print( new_seq )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 8<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;gmk Guanylate kinase<br>\n",
    "CGAATATTTGAAGATCCAAGTACATCATATAAGTATTCTATTTCAATGACAACACGTCAA<br>\n",
    "ATGCGTGAAGGTGAAGTTGATGGCGTAGATTACTTTTTTAAAACTAGGGATGCGTTTGAA<br>\n",
    "GCTTTAATCAAAGATGACCAATTTATAGAATATGCTGAATATGTAGGCAACTATTATGGT<br>\n",
    "ACACCAGTTCAATATGTTAAAGATACAATGGACGAAGGTCATGATGTATTTTTAGAAATT<br>\n",
    "GAAGTAGAAGGTGCAAAGCAAGTTAGAAAGAAATTTCCAGATGCGCTATTTATTTTCTTA<br>\n",
    "GCACCTCCAAGTTTAGAACACTTGAGAGAGCGATTAGTAGGTAGAGGAACAGAATCTGAT<br>\n",
    "GAGAAAATACAAAGTCGTATTAACGAAGCGCGTAAAGAAGTTGAAATGATGAATTTA\n",
    "</p>\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:10pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "and:\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&lt;class 'str'&gt;<br>\n",
    "C<br><br>\n",
    "&lt;class 'str'&gt;<br>\n",
    "ATATTT\n",
    "</p>\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:10pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "and:\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&lt;class 'str'&gt;<br>\n",
    "C<br><br>\n",
    "&lt;class '__main__.DNASeq'&gt;<br>\n",
    "&gt;gmk_loc(4_9) Guanylate kinase<br>\n",
    "ATATTT<br><br>\n",
    "&lt;class '__main__.DNASeq'&gt;<br>\n",
    "&gt;gmk_loc(1_1) Guanylate kinase<br>\n",
    "C\n",
    "</p>\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:10pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "and:\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;gmk_loc(4_9) Guanylate kinase<br>\n",
    "ATATTT<br><br>\n",
    "&gt;gmk_loc(4_9)_revcmpl Guanylate kinase<br>\n",
    "AAATAT<br><br>\n",
    "&gt;gmk_loc(4_9)_revcmpl Guanylate kinase<br>\n",
    "AAATAT\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gmk Guanylate kinase\n",
      "CGAATATTTGAAGATCCAAGTACATCATATAAGTATTCTATTTCAATGACAACACGTCAA\n",
      "ATGCGTGAAGGTGAAGTTGATGGCGTAGATTACTTTTTTAAAACTAGGGATGCGTTTGAA\n",
      "GCTTTAATCAAAGATGACCAATTTATAGAATATGCTGAATATGTAGGCAACTATTATGGT\n",
      "ACACCAGTTCAATATGTTAAAGATACAATGGACGAAGGTCATGATGTATTTTTAGAAATT\n",
      "GAAGTAGAAGGTGCAAAGCAAGTTAGAAAGAAATTTCCAGATGCGCTATTTATTTTCTTA\n",
      "GCACCTCCAAGTTTAGAACACTTGAGAGAGCGATTAGTAGGTAGAGGAACAGAATCTGAT\n",
      "GAGAAAATACAAAGTCGTATTAACGAAGCGCGTAAAGAAGTTGAAATGATGAATTTA\n"
     ]
    }
   ],
   "source": [
    "# przeładowanie, aby utworzyć obiekty na podstawie\n",
    "# najświeższej wersji klasy DNASeq\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "seq = seqs['gmk']\n",
    "\n",
    "print(seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "C\n",
      "\n",
      "<class 'str'>\n",
      "ATATTT\n"
     ]
    }
   ],
   "source": [
    "# indeksowanie w Pythonie zaczyna się od 0,\n",
    "# w przypadku slice'ów drugi indeks jest wyłączny,\n",
    "# możemy to zobaczyć operując bezpośrednio na sekwencji\n",
    "# obiektu typu DNASeq, która jest zmienną tekstową\n",
    "\n",
    "letter = seq.seq[0]\n",
    "\n",
    "print(type(letter), letter, sep='\\n', end='\\n\\n')\n",
    "\n",
    "s = seq.seq[3:9]\n",
    "\n",
    "print(type(s), s, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "C\n",
      "\n",
      "<class '__main__.DNASeq'>\n",
      ">gmk_loc(4_9) Guanylate kinase\n",
      "ATATTT\n",
      "\n",
      "<class '__main__.DNASeq'>\n",
      ">gmk_loc(1_1) Guanylate kinase\n",
      "C\n"
     ]
    }
   ],
   "source": [
    "# dzięki implementacji metody specjalnej __getitem__,\n",
    "# możemy indeksować obiekt typu DNASeq bezpośrednio, zgodnie\n",
    "# z notacją GenBank, czyli indeksując od 1 z drugim indeksem włącznie,\n",
    "# porównaj to z przykładem wyżej indeksowania w Pythonie\n",
    "\n",
    "letter = seq[1]\n",
    "\n",
    "print(type(letter), letter, sep='\\n', end='\\n\\n')\n",
    "\n",
    "s = seq[4:9]\n",
    "\n",
    "print(type(s), s, sep='\\n', end='\\n\\n')\n",
    "\n",
    "\n",
    "# aby otrzymać zmienną typu DNASeq zamiast str, przy pojedynczym znaku,\n",
    "# możemy utworzyć slice'a o długości 1, zgdonie z notacją GenBank\n",
    "\n",
    "s = seq[1:1]\n",
    "\n",
    "print(type(s), s, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gmk_loc(4_9) Guanylate kinase\n",
      "ATATTT\n",
      "\n",
      ">gmk_loc(4_9)_revcmpl Guanylate kinase\n",
      "AAATAT\n",
      "\n",
      ">gmk_loc(4_9)_revcmpl Guanylate kinase\n",
      "AAATAT\n"
     ]
    }
   ],
   "source": [
    "# ponieważ slice obiektu DNASeq zwraca te sam typ\n",
    "# obiektu możemy dalej pracować z fragmentem sekwencji\n",
    "# tak jak z całą sekwencją,\n",
    "# tutaj wyświetlacjąc jego sekwencję komplementarną\n",
    "# w formacie FASTA\n",
    "\n",
    "print(seq[4:9], end='\\n\\n')\n",
    "\n",
    "print(seq[4:9].revcmpl(), end='\\n\\n')\n",
    "\n",
    "# sekwencję komplementarną możemy wyświetlić,\n",
    "# korzystając z zaimplementowanego indeksowania\n",
    "# GenBank, podając odwrócone wartości indeksów\n",
    "\n",
    "print(seq[9:4])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 9<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;arcC_loc(1_10) Carbamate kinase<br>\n",
    "TTATTAATCC<br><br>\n",
    "&gt;glpF_loc(1_10) Glycerol kinase<br>\n",
    "GGTGCTGATT<br><br>\n",
    "&gt;arcC_loc(1_10)_glpF_loc(1_10) Carbamate kinase_Glycerol kinase<br>\n",
    "TTATTAATCCGGTGCTGATT\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">arcC_loc(1_10) Carbamate kinase\n",
      "TTATTAATCC\n",
      "\n",
      ">glpF_loc(1_10) Glycerol kinase\n",
      "GGTGCTGATT\n",
      "\n",
      ">arcC_loc(1_10)_glpF_loc(1_10) Carbamate kinase_Glycerol kinase\n",
      "TTATTAATCCGGTGCTGATT\n"
     ]
    }
   ],
   "source": [
    "# reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# select 10 nucleotide fragments of two different\n",
    "# sequences and add them together\n",
    "seq_1 = seqs['arcC'][1:10]\n",
    "\n",
    "seq_2 = seqs['glpF'][1:10]\n",
    "\n",
    "seq_3 = seq_1 + seq_2\n",
    "\n",
    "print(seq_1, seq_2, seq_3, sep='\\n\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 10<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "pta<br><br>\n",
    "new object<br><br>\n",
    "===<br><br>\n",
    "pta<br><br>\n",
    "new ref. the same obj.<br><br>\n",
    "new ref. the same obj.\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pta\n",
      "\n",
      "new object\n",
      "\n",
      "===\n",
      "\n",
      "pta\n",
      "\n",
      "new ref. the same obj.\n",
      "\n",
      "new ref. the same obj.\n"
     ]
    }
   ],
   "source": [
    "# reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# select one of the sequences by its sequence id (seqid)\n",
    "seq = seqs['pta']\n",
    "\n",
    "# our method copy() allows to create\n",
    "# an independent copy of an existing object\n",
    "\n",
    "seq_copy = seq.copy()\n",
    "\n",
    "# an object copy is an independent object,\n",
    "# separatly written in RAM memory\n",
    "seq_copy.seqid = 'new object'\n",
    "\n",
    "print(seq.seqid, seq_copy.seqid, sep='\\n\\n', end='\\n\\n===\\n\\n')\n",
    "\n",
    "# in contrast to the operation of assignment,\n",
    "# which simply creates a new reference to\n",
    "# the same, already existing in memory object\n",
    "\n",
    "print(seq.seqid, end='\\n\\n')\n",
    "\n",
    "new_ref = seq\n",
    "\n",
    "new_ref.seqid = 'new ref. the same obj.'\n",
    "\n",
    "print(seq.seqid, new_ref.seqid, sep='\\n\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddeeff;border-radius:15px\">The End :)</p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
